packages/atclient/src/connection.c

```diff
@@ -602,12 +578,32 @@ int atclient_connection_read(atclient_connection *ctx, unsigned char **value, si
   /*
    * 4. Read the value
    */
-  ret = atclient_tls_socket_read(&ctx->_socket, value, value_len, atclient_socket_read_until_char('\n'));
+  unsigned char *read_buf;
+  size_t read_n;
+  ret = atclient_tls_socket_read(&ctx->_socket, &read_buf, &read_n, atclient_socket_read_until_char('\n'));
   if (ret != 0) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to read from the connection\n", ret);
     goto exit;
   }
-  *value[*value_len - 1] = '\0'; // replace '\n' with '\0'
+  size_t read_i = 0; // will store where the start of `<type>:` is (if happy path)
+  ret = atclient_utils_find_index_past_at_prompt(*value, *value_len, &read_i);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to parse the result read from the connection\n");
+    free(read_buf);
+    goto exit;
+  }
+  read_n -= read_i;
+  *value = malloc(read_n * sizeof(char));
+  if (*value == NULL) {
+    free(read_buf);
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to allocate memory for the final output buffer\n");
+    goto exit;
+  }
+  memcpy(*value, read_buf + read_i, read_n);
+  free(read_buf);
+  *value[read_n - 1] = '\0'; // null terminate the string
+  *value_len = read_n;
+
```

Comment:
make sure that we strip prompts in atclient_connect_read, previously this was only done in atclient_connection_send.

Also ensures that the original read buffer is freed in full, so we need a new allocation here. I know how to remove this extra allocation later, but for now, this is the quickest path to robust connections without redesigning how memory management is handled.